MAINT: Fix a dtype check in `scipy.signal._waveforms.py`

pavyamsiri opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes #22252.

#### What does this implement/fix?
<!--Please explain your changes.-->
The old code used to call `dtype.char` and then check if it was contained with a list of a single string `['fdFD']`. This of course would never pass because a `'f'` is not a member of `['fdFD']` but it is a member of `'fdFD'`.

This is likely a mistake in the code and so this commit fixes it.

pavyamsiri committed:
MAINT: Fix dtype check in `scipy.signal._waveforms`.

The old code used to call `dtype.char` and then check if it was
contained with a list of a single string `['fdFD']`. This of course
would never pass because a `'f'` is not a member of `['fdFD']` but it is
a member of `'fdFD'`.

This is likely a mistake in the code and so this commit fixes it.

j-bowhay commented:
Thanks for spotting this @pavyamsiri ! It would be great if you could also add a test that previously failed because of this bug

pavyamsiri committed:
TST: Add tests for dtypes for the functions `square` and `sawtooth`

In `scipy.signal`.

Before the previous commit, these tests would fail as they always
defaulted to `np.float64`, even if the input parameters were
`np.float32`.

nickodell commented:
> I'm not entirely sure why the CI fails though? My initial PR changed a single file and it caused docs to fail build?

The docs build failure looks unrelated. You can ignore that. See #22249 for details.

However, the test added in this PR is not passing on all platforms. For example, it is failing for [32-bit, fast, py3.10/npMin, dev.py](https://github.com/scipy/scipy/actions/runs/12632657374/job/35196769214#logs). It doesn't appear to happen on all platforms. Can you investigate this?

```python

=================================== FAILURES ===================================
_______________________ TestSawtoothWaveform.test_dtype ________________________
scipy/signal/tests/test_waveforms.py:385: in test_dtype
    waveform = waveforms.sawtooth(
        self       = <scipy.signal.tests.test_waveforms.TestSawtoothWaveform object at 0xd8c14778>
scipy/signal/_waveforms.py:76: in sawtooth
    place(y, mask2, tsub / (pi * wsub) - 1)
        mask1      = False
        mask2      = 1
        t          = array(1., dtype=float32)
        tmod       = 1.0
        tsub       = array([1.])
        w          = array(1., dtype=float32)
        width      = 1.0
        wsub       = array([1.], dtype=float32)
        y          = array(0., dtype=float32)
        ytype      = 'f'
<__array_function__ internals>:180: in place
    ???
        args       = (array(0., dtype=float32), 1, array([-0.68169012]))
        dispatcher = <function _place_dispatcher at 0xe4b57928>
        implementation = <function place at 0xe4b57a48>
        kwargs     = {}
        public_api = <function place at 0xe4b57a90>
        relevant_args = (array(0., dtype=float32), 1, array([-0.68169012]))
/scipy/test/lib/python3.10/site-packages/numpy/lib/function_base.py:1953: in place
    return _insert(arr, mask, vals)
E   TypeError: Cannot cast array data from dtype('float64') to dtype('float32') according to the rule 'safe'
        arr        = array(0., dtype=float32)
        mask       = 1
        vals       = array([-0.68169012])
=========================== short test summary info ============================
FAILED scipy/signal/tests/test_waveforms.py::TestSawtoothWaveform::test_dtype
= 1 failed, 59229 passed, 3591 skipped, 16092 deselected, 595 xfailed, 12 xpassed in 771.30s (0:12:51) =
```

ev-br commented:
It's a curious bug indeed. That said, a bug which was there for two decades is quite likely a feature by now. All users of these functions were always getting float64 results for all inputs. Therefore, what you do in this PR is actually an enhancement, which, as you said, might have a backwards compat effect.

So I'd suggest to at least consider being explicit in always returning float64 waveforms. Increasing test for functions which were not covered previously is of course most welcome though.

pavyamsiri commented:
> It's a curious bug indeed. That said, a bug which was there for two decades is quite likely a feature by now. All users of these functions were always getting float64 results for all inputs. Therefore, what you do in this PR is actually an enhancement, which, as you said, might have a backwards compat effect.
> 
> So I'd suggest to at least consider being explicit in always returning float64 waveforms. Increasing test for functions which were not covered previously is of course most welcome though.

Do you mean that we should instead remove the dtype checks entirely and make it so that it always returns `float64`? I don't mind doing this.

As you said this bug has been around for 2 decades so clearly the intended functionality is not that important for users.

pavyamsiri commented:
> > I'm not entirely sure why the CI fails though? My initial PR changed a single file and it caused docs to fail build?
> 
> The docs build failure looks unrelated. You can ignore that. See #22249 for details.
> 
> However, the test added in this PR is not passing on all platforms. For example, it is failing for [32-bit, fast, py3.10/npMin, dev.py](https://github.com/scipy/scipy/actions/runs/12632657374/job/35196769214#logs). It doesn't appear to happen on all platforms. Can you investigate this?
> 
> ```python
> =================================== FAILURES ===================================
> _______________________ TestSawtoothWaveform.test_dtype ________________________
> scipy/signal/tests/test_waveforms.py:385: in test_dtype
>     waveform = waveforms.sawtooth(
>         self       = <scipy.signal.tests.test_waveforms.TestSawtoothWaveform object at 0xd8c14778>
> scipy/signal/_waveforms.py:76: in sawtooth
>     place(y, mask2, tsub / (pi * wsub) - 1)
>         mask1      = False
>         mask2      = 1
>         t          = array(1., dtype=float32)
>         tmod       = 1.0
>         tsub       = array([1.])
>         w          = array(1., dtype=float32)
>         width      = 1.0
>         wsub       = array([1.], dtype=float32)
>         y          = array(0., dtype=float32)
>         ytype      = 'f'
> <__array_function__ internals>:180: in place
>     ???
>         args       = (array(0., dtype=float32), 1, array([-0.68169012]))
>         dispatcher = <function _place_dispatcher at 0xe4b57928>
>         implementation = <function place at 0xe4b57a48>
>         kwargs     = {}
>         public_api = <function place at 0xe4b57a90>
>         relevant_args = (array(0., dtype=float32), 1, array([-0.68169012]))
> /scipy/test/lib/python3.10/site-packages/numpy/lib/function_base.py:1953: in place
>     return _insert(arr, mask, vals)
> E   TypeError: Cannot cast array data from dtype('float64') to dtype('float32') according to the rule 'safe'
>         arr        = array(0., dtype=float32)
>         mask       = 1
>         vals       = array([-0.68169012])
> =========================== short test summary info ============================
> FAILED scipy/signal/tests/test_waveforms.py::TestSawtoothWaveform::test_dtype
> = 1 failed, 59229 passed, 3591 skipped, 16092 deselected, 595 xfailed, 12 xpassed in 771.30s (0:12:51) =
> ```

That seems very strange. It seems that numpy's `place` function is using `float64` for that platform and then because the inputs are `float32` it can't do a conversion.

I don't know how to test that platform without just running the CI again. Do you know of a way to test that platform?

pavyamsiri commented:
I have figured out the cause of the bug.

Prior to numpy 2, the type promotion rules caused this code to output `float64` no matter the input array dtype.

```python
tmod = mod(t, 2 * pi)
```

This is because the divisor is a python `float` which at the precision of `pi` becomes `float64`. In numpy 2, type promotion rules were updated to better take into account the input dtypes (see [NEP 50](https://numpy.org/neps/nep-0050-scalar-promotion.html#nep50)).

I could fix this by recasting the promoted arrays back into the input dtypes, but I'm not sure if want to continue with adding this enhancement or going back to old behaviour and just removing the always false check as @ev-br suggested (assuming I understood his comment correctly).

pavyamsiri committed:
BUG: Add explicit recasts of intermediate arrays in `sawtooth`

Before this commit, the `sawtooth` function in
`scipy.signal._waveforms.py` could error out on numpy v1.x due to a
change in type promotion rules between v1 and v2.

Prior to numpy 2, type promotion was value based meaning the intermediate
values could be promoted from `float32` to `float64` due to use of
python `float` scalars, which then promotes the output dtype to
`float64`, which leads to errors as the `place` function uses safe type
casting.

This patch avoid this by casting the intermediate arrays back to the
output dtype for each call of `place`.

jorenham commented on scipy/signal/_waveforms.py:
```diff
@@ -73,15 +73,17 @@ def sawtooth(t, width=1):
     tsub = extract(mask2, tmod)
     wsub = extract(mask2, w)
-    place(y, mask2, tsub / (pi * wsub) - 1)
+    values = (tsub / (pi * wsub) - 1)
```
```suggestion
    values = tsub / (pi * wsub) - 1
```

pavyamsiri replied:
NIce catch!

pavyamsiri committed:
MAINT: Remove unnecessary parentheses

Under `sawtooth` in `scipy.signal._waveforms.py`.

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

pavyamsiri commented:
I'm not sure why the CI is failing here? The tests seem to fail for a completely unrelated test to this PR (in `scipy.stats`). Is this due to another issue?

pavyamsiri commented:
@ev-br Sorry to bother you if you are busy, just wondering on your thoughts on this PR still? I'm not entirely sure we should change the behaviour in order to fix the bug or remove the dead code entirely and return float64 waveforms only.

ev-br commented:
FWIW, I'd suggest going with a backwards compatible option to always return f64 wafeforms.

pavyamsiri commented:
> FWIW, I'd suggest going with a backwards compatible option to always return f64 wafeforms.

Okay I will probably just make a different PR that removes the dead code but keeps the current functionality then.

ev-br commented:
superseded by https://github.com/scipy/scipy/pull/22520. Thank you @pavyamsiri, for both PRs!
